apps/builder/app/builder/shared/css-editor/css-editor.tsx

```diff
@@ -0,0 +1,504 @@
+    }
+
+    const controller = new AbortController();
+
+    ref.current.addEventListener(
+      "contentvisibilityautostatechange",
+      (event) => {
+        setIsVisible(!event.skipped);
+      },
+      {
+        signal: controller.signal,
+      }
+    );
+
+    return () => {
+      controller.abort();
+    };
+  }, [visibilityChangeEventSupported]);
+
+  return (
+    <div
+      ref={ref}
+      style={{
+        display: "inline-block",
+        contentVisibility: "auto",
+        // https://developer.mozilla.org/en-US/docs/Web/CSS/contain-intrinsic-size
+        // containIntrinsicSize is used to set the default size of an element before any content is loaded.
+        // This helps in preventing layout shifts and provides a better user experience by maintaining a consistent layout.
+        // It also affects the contentvisibilityautostatechange event to be called properly,
+        // with "auto" it will call it with skipped false for all initial elements.
+        // 44px is the height of the property row with 2 lines of text. This value can be adjusted slightly.
+        containIntrinsicSize: "auto 44px",
+      }}
+    >
+      {isVisible ? children : undefined}
+    </div>
+  );
+};
+
+const AdvancedDeclarationLonghand = memo(
+  ({
+    property,
+    onChangeComplete,
+    deleteProperty,
+    setProperty,
+    onReset,
+    valueInputRef,
+    indentation = initialIndentation,
+  }: {
+    property: CssProperty;
+    indentation?: string;
+    onReset?: () => void;
+    setProperty: SetProperty;
+    deleteProperty: DeleteProperty;
+    onChangeComplete?: ComponentProps<
+      typeof CssValueInputContainer
+    >["onChangeComplete"];
+    valueInputRef?: RefObject<HTMLInputElement>;
+  }) => {
+    return (
+      <Flex
+        css={{ paddingLeft: indentation }}
+        wrap="wrap"
+        align="center"
+        justify="start"
+        {...{ [copyAttribute]: property }}
+      >
+        <AdvancedPropertyLabel
+          property={property}
+          onReset={onReset}
+          deleteProperty={deleteProperty}
+        />
+        <Text
+          variant="mono"
+          // Improves the visual separation of value from the property.
+          css={{
+            textIndent: "-0.5ch",
+            fontWeight: "bold",
+          }}
+        >
+          :
+        </Text>
+        <AdvancedPropertyValue
+          property={property}
+          onChangeComplete={onChangeComplete}
+          onReset={onReset}
+          deleteProperty={deleteProperty}
+          setProperty={setProperty}
+          inputRef={valueInputRef}
+        />
+      </Flex>
+    );
+  }
+);
+
+export type CssEditorApi = { showAddStyleInput: () => void } | undefined;
+
+export const CssEditor = ({
+  deleteProperty,
+  setProperty,
+  addProperties,
+  styleMap,
+  apiRef,
+  showSearch = true,
+  recentProperties = [],
+}: {
+  deleteProperty: DeleteProperty;
+  setProperty: SetProperty;
+  addProperties: (styleMap: StyleMap) => void;
```

Comment:
Maybe use callback style like onCreate, onChange, onDelete?

Reply:
Ah I know why, because we don't use these functions anywhere as callbacks, these are global imports and only here we needed to abstract away from these.

Reply:
same thing in CssValueInputContainer